063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,271 @@
+
+1. Request pending replicas changes (uses `topic_configuration` endpoint, and returns topics with pending and ongoing changes).
+2. Check ongoing replicas changes (uses `user_tasks` endpoint, and returns topics with completed changes).
+
+The result of these operations is used to update the KafkaTopic status.
+
+### Error handling
+
+When there is an error, the replicas change remains pending or ongoing (no status update), and it is indefinitely retried by the periodic reconciliation.
+All errors are written in the Topic Operator logs, so that the user can verify the reason of a never ending replicas change.
+
+In Cruise Control, there is no way to differentiate between temporary errors (resolve by themself) from permanent errors (require some action).
+For example, the `NotEnoughValidWindowsException` can be raised when Cruise Control has just started and the cluster model is still building, but also when there is a configuration error which prevents broker metrics collection.
+Another example is the `OptimizationFailureException` raised when some hard goal is violated, which can be temporary if the configured network capacity is less than the real network capacity and there is a traffic peak.
+Instead, the `OngoingExecutionException` is clearly temporary and it means that there is already a task that is being executed by Cruise Control.
+
+Cruise Control allows to scale down the replication factor under the `min.insync.replicas` value, and this can cause disruption to producers with `acks=all`.
+When this happens, the Topic Operator logs a warning but continues with the replicas change, because the KafkaRoller ignores topics with RF < minISR, and they don't event show up as under replicated.
+The target replication factor should also be less than or equal to the number of available brokers, but this is enforced directly by Cruise Control.
+
+When a manged KafkaTopic with replicas change is deleted in Kubernetes, the Topic Operator also deletes the topic in Kafka, but the task execution continues in Cruise Control.
+There is little benefit in using the `stop_proposal_execution` endpoint, because most replicas change tasks will have only one batch that can't be stopped.
+
+When a manged topic with replicas change is deleted in Kafka, the Topic Operator recreates the topic with the target replication factor.
+
+When Topic Operator is restarted, both pending and ongoing replicas changes are recovered from KafkaTopic status.
+The ongoing state includes the `userTaskId`, which is the unique identifier of a Cruise Control task, which includes one or more replicas changes.
+
+When Cruise Control is restarted, all active tasks are lost, as there is no persistent memory.
+The Topic Operator detects this event when the ongoing replicas changes check returns no tasks for a known `userTaskId`.
+All pending and ongoing replicas changes are still stored in KafkaTopic status, and can be recovered as soon as Cruise Control becomes Ready.
+
```

Comment:
You said before that CC allows only a fixed number of pending tasks, so I'd like to see some discussion of how the TO handles that situation, including show it's reflected in the `KafkaTopic`.

Reply:
Sure, I added that part. It's the same error handling procedure. A clear message is surfaced to the user (There are already 5 active user tasks, which has reached the servlet capacity.), and the Topic Operator periodically retries until there is room for it.